### PR TITLE
refactor: use grr::matches do match data to samples

### PR DESCRIPTION
**Pull request by jorainer**
Use the `grr::matches` function instead of an `lapply` to find all matches between elements and samples in `linkSampleData`. This increases performance, especially for larger data sets.

The code below simulates creation of the *mapping matrix* that links rows in `sampleData` (i.e. samples) to elements in `spectra`. This first example is equivalent with an experiment with 100 files each having 100 spectra in it.

```r
library(microbenchmark)
library(MsExperiment)
> a <- 1:100
> b <- rep(1:100, each = 100)
> b <- rep(1:100, each = 100)
> microbenchmark(
+ MsExperiment:::.link_matrix(a, b),
+ MsExperiment:::.link_matrix2(a, b),
+ times = 17)
> 
Unit: microseconds
                               expr      min       lq      mean   median
  MsExperiment:::.link_matrix(a, b) 1920.169 2008.577 2716.4142 2387.868
 MsExperiment:::.link_matrix2(a, b)  530.839  594.414  640.6648  633.686
       uq      max neval cld
 2841.082 4751.933    17   b
  694.379  731.611    17  a 
```

The original `lapply`-based approach is slower, but it's not yet problematic. For larger data set this really becomes an issue: example 5000 files, each with 100 spectra:

```r
> a <- 1:5000
> b <- rep(1:5000, each = 100)
> microbenchmark(
+ MsExperiment:::.link_matrix(a, b),
+ MsExperiment:::.link_matrix2(a, b),
+ times = 17)
> 
Unit: milliseconds
                               expr        min         lq       mean    median
  MsExperiment:::.link_matrix(a, b) 5979.95425 6382.44017 6590.96746 6595.0389
 MsExperiment:::.link_matrix2(a, b)   29.79996   31.40307   51.03011   33.1364
         uq       max neval cld
 6791.41652 7237.3885    17   b
   38.21319  324.0143    17  a 
```

Get's even worse for larger data sets:

```r
> a <- 1:5000
> b <- rep(1:5000, each = 2000)
> microbenchmark(
MsExperiment:::.link_matrix(a, b),
MsExperiment:::.link_matrix2(a, b),
times = 17)
Unit: milliseconds
                               expr        min          lq       mean
  MsExperiment:::.link_matrix(a, b) 109202.110 112255.9484 118694.336
 MsExperiment:::.link_matrix2(a, b)    713.088    823.0377   1000.542
      median         uq        max neval cld
 117076.9932 123866.259 131894.952    17   b
    886.4045   1105.617   1622.613    17  a 

```

So, the original `lapply`-based approach indeed becomes an issue for larger data sets.

### CODE REVIEW

**Comment by sgibb** [comment]
Though `grr` is from 2016 I never heard about it. Looks impressive. Did you try `S4Vector::findMatches` as well?

**Comment by jorainer** [comment]
Thanks @sgibb , was not aware of `findMatches`! Comparison with that:

```r
library(S4Vectors)
.link_matrix3 <- function(from = integer(), to = integer()) {
    unname(as.matrix(findMatches(from, to)))
}

a <- 1:5000
b <- rep(1:5000, each = 100)
microbenchmark(
    MsExperiment:::.link_matrix(a, b),
    MsExperiment:::.link_matrix2(a, b),
    .link_matrix3(a, b),
    times = 17
)
Unit: milliseconds
                               expr        min         lq       mean     median
  MsExperiment:::.link_matrix(a, b) 5665.07333 5801.86135 5968.02212 5870.57092
 MsExperiment:::.link_matrix2(a, b)   24.50225   27.16140   29.90907   28.77457
                .link_matrix3(a, b)   19.96335   22.80254   24.68690   23.52615
         uq        max neval cld
 6203.50092 6331.94260    17   b
   31.53121   37.02756    17  a 
   24.43055   35.06611    17  a 

```

And for the larger data set:

```r
a <- 1:5000
b <- rep(1:5000, each = 2000)
microbenchmark(
    MsExperiment:::.link_matrix(a, b),
    MsExperiment:::.link_matrix2(a, b),
    .link_matrix3(a, b),
    times = 17
)
Unit: milliseconds
                               expr         min          lq        mean
  MsExperiment:::.link_matrix(a, b) 103536.4444 107822.5091 111105.5599
 MsExperiment:::.link_matrix2(a, b)    448.1719    726.0769    779.8090
                .link_matrix3(a, b)    521.2230    812.4368    950.7604
      median          uq        max neval cld
 112320.3945 114206.5389 116151.355    17   b
    749.4757    776.5993   1868.314    17  a 
    862.0807   1112.7778   1443.515    17  a 
```

So, `findMatches` and `grr::matches` have equivalent performance. I suggest to use `findMatches` because a) we already import from `S4Vectors` and could reduce thus the number of dependencies of the package and b) `findMatches` might be the more robust solution since `grr` was not updated in the last couple of years.

**Comment by jorainer** [comment]
Changed now from `grr::matches` to `S4Vectors::findMatches`.

File: NEWS.md [hunk 1]
```diff
@@ -1,3 +1,15 @@
+# MsExperiment 1.1
+
+## MsExperiment 1.1.1
+
+- Use `grr::matches` in `linkSampleData` which improves performance,
```
Comment (sgibb) [inline NEWS.md]:
```suggestion
- Use `S4Vectors::findMatches` in `linkSampleData` which improves performance,
```

File: R/MsExperiment-functions.R [hunk 2]
```diff
@@ -175,7 +175,15 @@
...
     res <- lapply(from, function(x) which(to == x))
     ls <- lengths(res)
-    cbind(rep(seq_along(from), ls), unlist(res[ls > 0], use.names = FALSE))
+    cbind(rep(seq_along(from), ls), unlist(res, use.names = FALSE))
+}
+
+#' @importFrom S4Vectors findMatches
+#'
+#' @noRd
+.link_matrix2 <- function(from = integer(), to = integer()) {
```
Comment (sgibb) [inline R/MsExperiment-functions.R]: Should become `link_matrix` and the old `link_matrix` should be removed.

File: R/MsExperiment.R [hunk 3]
```diff
@@ -410,8 +410,8 @@ linkSampleData <- function(object, with = character(),
             to_slot <- link_string[1L]
             with <- paste0(link_string[c(1, 2)], collapse = ".")
         } else stop("one of the slot names has to be 'sampleData'.")
-        link <- .link_matrix(.get_element(object, from),
-                             .get_element(object, with))
+        link <- .link_matrix2(.get_element(object, from),
```
Comment (sgibb) [inline R/MsExperiment.R]: Should not be necessary.

File: tests/testthat/test_MsExperiment-functions.R [hunk 4]
```diff
@@ -97,6 +97,27 @@ test_that(".link_matrix works", {
     expect_equal(res, cbind(c(1, 2, 3, 5), c(2, 2, 3, 3)))
 })
 
+test_that(".link_matrix2 works", {
```
Comment (sgibb) [inline tests/testthat/test_MsExperiment-functions.R]: Should be integrated into the original `link_matrix` tests.

**Review by sgibb — APPROVED** [review]
Nice! You should just replace the original `link_matrix` with `link_matrix2` (to avoid two functions/old code).

File: R/MsExperiment-functions.R (hunk 2 above)
Comment (jorainer) [reply R/MsExperiment-functions.R]: Done

File: tests/testthat/test_MsExperiment-functions.R (hunk 4 above)
Comment (jorainer) [reply tests/testthat/test_MsExperiment-functions.R]: Done

**Pull request merged**